### PR TITLE
VZ-7478 Add prerequisite validator webhook

### PR DESCRIPTION
**Pull request by ayaranga**
A webhook has been added to check the Verrazzano prerequisites and present an appropriate warning to the user while applying the Verrazzano custom resource.

### CODE REVIEW

File: pkg/k8sutil/k8sutil.go [hunk 1]
```diff
@@ -154,6 +154,24 @@ func GetKubernetesClientsetWithConfig(config *rest.Config) (*kubernetes.Clientse
 	return clientset, err
 }
 
+// GetKubernetesNodeList returns a list of Kubernetes nodes
```
Comment (pfmackin) [inline pkg/k8sutil/k8sutil.go:157]: put this code in a new package/file   verrazzzano/pkg/k8s/node/node.go

File: pkg/vzchecks/requirement.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		ephemeralStorage: &resourceInfo{resourceType: EphemeralStorage, allocatable: resource.MustParse("100G")},
+	}
+	ProdProfile = VZRequirement{
+		nodeCount:        3,
+		cpu:              &resourceInfo{resourceType: CPU, allocatable: resource.MustParse("4")},
+		memory:           &resourceInfo{resourceType: Memory, allocatable: resource.MustParse("32G")},
+		ephemeralStorage: &resourceInfo{resourceType: EphemeralStorage, allocatable: resource.MustParse("100G")},
+	}
+	ManagedProfile = VZRequirement{
+		nodeCount:        3,
```
Comment (pfmackin) [inline pkg/vzchecks/requirement.go:64]: I think this is too high for managed cluster.  Customers might have edge clusters that are very lightweight single node.  Count should be 1, but double check with Kevin.  Also, memory seems high but maybe it is ok.

File: pkg/k8sutil/k8sutil.go (hunk 1 above)
Comment (ayaranga) [reply pkg/k8sutil/k8sutil.go:157]: Done. Thanks.

File: pkg/vzchecks/requirement.go (hunk 2 above)
Comment (ayaranga) [reply pkg/vzchecks/requirement.go:64]: I have changed it to 1 and checking with Kevin as well.

File: pkg/vzchecks/requirement.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		cpu:              &resourceInfo{resourceType: CPU, allocatable: resource.MustParse("4")},
+		memory:           &resourceInfo{resourceType: Memory, allocatable: resource.MustParse("32G")},
+		ephemeralStorage: &resourceInfo{resourceType: EphemeralStorage, allocatable: resource.MustParse("100G")},
+	}
+	profileMap = map[ProfileType]VZRequirement{
+		Dev:            DevProfile,
+		Prod:           ProdProfile,
+		ManagedCluster: ManagedProfile,
+	}
+)
```
[the remark this answers is not in the document]
Comment (ayaranga) [reply pkg/vzchecks/requirement.go:74]: I had defaulted to prod in precheck file. Made a change to check in this file as well.

File: pkg/vzchecks/requirement.go (hunk 2 above)
Comment (kminder) [reply pkg/vzchecks/requirement.go]: Should be 1.

File: pkg/vzchecks/precheck.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	clipkg "sigs.k8s.io/controller-runtime/pkg/client"
+	"strconv"
+)
+
+// PrerequisiteCheck checks the prerequisites before applying the Verrazzano CR
+func PrerequisiteCheck(client clipkg.Client, profile ProfileType) []error {
+	return preCheck(client, profile)
+}
+
+func preCheck(client clipkg.Client, profile ProfileType) []error {
```
Comment (kminder) [inline pkg/vzchecks/precheck.go:20]: This probably shouldn't be using profile as the primary means to determine what the resource requirements are going to be.  We should discuss.

File: pkg/vzchecks/requirement.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+const (
+	CPU              = ResourceType(k8score.ResourceCPU)
+	Memory           = ResourceType(k8score.ResourceMemory)
+	EphemeralStorage = ResourceType(k8score.ResourceEphemeralStorage)
+)
+
+const (
+	nodeCountReqMsg = "minimum required number of worker nodes for %s profile is %d but the available number of worker nodes is %d"
```
Comment (kminder) [inline pkg/vzchecks/requirement.go:44]: At a minimum take the profile statement out of these messages.

Comment (ayaranga) [reply pkg/vzchecks/requirement.go]: Done

File: pkg/vzchecks/precheck.go (hunk 4 above)
Comment (ayaranga) [reply pkg/vzchecks/precheck.go:20]: As discussed, we will take this up as a separate incremental PR

**Review by kminder — APPROVED** [review]
Approved as an intermediate solution.  This needs to be followed up such that the requirements are derived from the enabled components and their configuration rather than relying on the base profile name alone.

**Pull request merged**